src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Analyzer.Utilities;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    // TODO: VisualBasic as well?
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class StringBuilderAppendShouldNotTakeSubstringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray
+            .Create(
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId,
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdTwoParameterId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            SyntaxNode node = root.FindNode(context.Span);
+
+            // TODO is it guaranteed that there's only one at once?
+            Diagnostic diagnostic = context.Diagnostics.Single();
```

Comment:
I usually loop but have no idea. Weird API if they use a collection if there is a guarantee there is only ever one.

Reply:
;) that's why I did it exactly this way, including the TODO - wanted to know what's the case, as e.g. in  AbstractTypesShouldNotHavePublicConstructorsFixer, DefineAccessorsForAttributeArgumentsFixer, EnumsShouldHaveZeroValueFixer and more.

fixing now, but perhaps the other fixers should be looked into then

Reply:
See this comment for details: https://github.com/dotnet/roslyn/pull/30093#issuecomment-424145149

Gist:
```
In short, majority of fixers should take the following approach:
1. Operate on the first diagnostic from the code fix context.
2. Have defensive checks for the syntax kind to which a diagnostic span maps, and bail out for unexpected cases.
3. Use the first diagnostic for RegisterCodeFix invocation.
```

Reply:
looping now, so this should be fixed.

Reply:
i don't see the looping :)

Reply:
sorry @CyrusNajmabadi that was not yet pushed, now it is.